### PR TITLE
Moving json to internal

### PR DESCRIPTION
**Pull request by vhvb1989**
fixes: #1377

### CODE REVIEW

File: sdk/identity/azure-identity/CHANGELOG.md [hunk 1]
```diff
@@ -1,8 +1,5 @@
 # Release History
 
-## 1.0.0-beta.3 (Unreleased)
-
-
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/CHANGELOG.md:5]: Revert this change. Why remove it?

File: sdk/core/azure-core/CMakeLists.txt [hunk 2]
```diff
@@ -32,17 +32,6 @@ if(BUILD_TRANSPORT_CURL)
   message("Libcurl version ${CURL_VERSION_STRING}")
 endif()
 
-# Storage requires 3.6.0 version for using `contains` feature
-set(NLOHMANN_JSON_MIN_REQUIRED_VERSION 3.6.0)
```
Comment (ahsonkhan) [inline sdk/core/azure-core/CMakeLists.txt:36]: Add a comment in source somewhere or persist some other indication which lets us remember that we we picked version `3.6.0` of nlohmann json as the source copy, so we know what to update from/to if need be in the future.

File: sdk/identity/azure-identity/CHANGELOG.md (hunk 1 above)
Comment (vhvb1989) [reply sdk/identity/azure-identity/CHANGELOG.md:5]: My bad.  Weird. THanks

**Comment by Jinming-Hu** [comment]
@vhvb1989 Who do we need to do this? Why doesn't the old way of installing it via vcpkg work for us?

**Review by Jinming-Hu — CHANGES_REQUESTED** [review]
I don't think it's the ideal way to include another library directly in this repo. We'll lose the benefit of getting updates including new features and bug fixes.

Is there any good reason of doing so?

**Comment by vhvb1989** [comment]
> Is there any good reason of doing so?

@Jinming-Hu 

Let me add some details about the Json story from the start to the current and proposed solution.

- First Approach
Json lib header was part of storage package only and exposed from public headers.
Cons:
- Can't re-use the same json lib from another service.
- Can't use a copy of the json lib in another service since it would collapse with the one in Storage when linking.

- Second approach
Moved the json header to storage. This solved the first cons but there is still other cons:
- End-customer collision if they are already using the json lib

- Third approach (current)
Made json lib an external dependency. It needs to be installed, for example,  with vcpkg. This solved the problem of collision with another copy of the lib when integrating to another solution (end-user code or some other 3rd party lib). 
But now the cons:
- Min version need to be followed. If end-users or some other lib is using an older version, they need to update in order to use our SDK

- Proposed
Json lib header is part of Azure Core but it is some how hidden under Azure::Core::Interal namespace.  This prevents collision and allow 2 versions of the json lib living together.
This also break the pre-install dependency and min version mandate.
When it comes to updating to a new version, it becomes invisible to end-customers. We don't need to ask them to update the json lib version as well.  We would just replace the json header we have and make some changes to it so it is under out internal namespace.

As you can see, all previous versioning/dependency/collision issues are gone.  This approach is simulating that Azure Core is providing its own Json API by borrowing a 3rd party implementation.

**Comment by Jinming-Hu** [comment]
@vhvb1989 So the only con of current approach is that it requires a minimum version. I've been thinking such a requirement is a bad decision, made unintentionally. If we remove such a requirement, do you think the current approach will be better than what you proposed in this PR?

**Comment by JeffreyRichter** [comment]
I'll add that we won't always want the latest version which may have bug fixes, but it may also introduce bugs and security issues. In addition, a newer version may not be backward compatible, and this could break us. So by embedded in dependency privately, we are in complete control of this dependency AND we don't conflict with a customer's desire to depend on a different version of this dependency.

**Review by Jinming-Hu — APPROVED** [review]
> I'll add that we won't always want the latest version which may have bug fixes, but it may also introduce bugs and security issues. In addition, a newer version may not be backward compatible, and this could break us. So by embedded in dependency privately, we are in complete control of this dependency AND we don't conflict with a customer's desire to depend on a different version of this dependency.

okay

**Review by RickWinter — COMMENTED** [review]
Where are the json tests in our tree?  We need a full suite since we are incorporating the code in our branch

**Comment by ahsonkhan** [comment]
> Where are the json tests in our tree? We need a full suite since we are incorporating the code in our branch

Yes please. We need a testing strategy. Thanks for brining it up Rick!

**Comment by Jinming-Hu** [comment]
image

What happened?

**Comment by vhvb1989** [comment]
> 
> 
> image
> 
> What happened?
> 
> 1300+ files changed and 6 millions lines of code added?

I am working to add the nlohmann unit tests to Core as well.  The 3.6 includes all the test data as part of the repo, that's why it comes with millions of lines of code for the test data.  
I have moved to 3.8 where the test data is fetched from external source instead of having it in the repo

**Comment by Jinming-Hu** [comment]
> > image
> > What happened?
> > 1300+ files changed and 6 millions lines of code added?
> 
> I am working to add the nlohmann unit tests to Core as well. The 3.6 includes all the test data as part of the repo, that's why it comes with millions of lines of code for the test data.
> I have moved to 3.8 where the test data is fetched from external source instead of having it in the repo

Do we really have to do that?

**Review by Jinming-Hu — CHANGES_REQUESTED** [review]
Storage team has big concern, do not merge

**Review by Jinming-Hu — APPROVED** [review]

File: sdk/core/azure-core/CMakeLists.txt (hunk 2 above)
Comment (vhvb1989) [reply sdk/core/azure-core/CMakeLists.txt:36]: the version is in the json.hpp file

**Comment by vhvb1989** [comment]
> 
> 
> Where are the json tests in our tree? We need a full suite since we are incorporating the code in our branch

Will create another PR for it

File: sdk/core/azure-core/vcpkg/CONTROL [hunk 3]
```diff
@@ -3,7 +3,7 @@
 #
 Source: azure-core-cpp
 Version: @AZ_LIBRARY_VERSION@
-Build-Depends: curl, nlohmann-json
```
Comment (antkmsft) [inline sdk/core/azure-core/vcpkg/CONTROL:6]: 👍

File: sdk/core/azure-core/vcpkg/Config.cmake.in [hunk 4]
```diff
@@ -5,7 +5,6 @@
 
 include(CMakeFindDependencyMacro)
 find_dependency(Threads)
-find_dependency(nlohmann_json @NLOHMANN_JSON_MIN_REQUIRED_VERSION@)
```
Comment (antkmsft) [inline sdk/core/azure-core/vcpkg/Config.cmake.in:8]: 👍

**Review by antkmsft — APPROVED** [review]

**Comment by Jinming-Hu** [comment]
Well, I'm afraid i'm gonna propose to revisit the decision of adding nlohmann json source code into this repo if you insist adding the huge test suite. What's on earth the benefit of doing so? given we'll remove the minimum version requirement in the near future? Whatever your argument is, I guess it will also apply to other dependencies. So my question is, why don't we also add libcurl source code into this repo? What's the difference between libcurl and nlohmann json to us?

**Review by antkmsft — CHANGES_REQUESTED** [review]
The diff is not visible because the file is too large, but we're adding ~2000 lines worth of macro definitions alone, none of them are AZ_ prefixed, most can be removed, and all would collide with NLOHMANN_JSON. Please remove most of them that can be removed, AZ_ prefix the rest.

**Comment by antkmsft** [comment]
@vhvb1989, I am talking about these: https://github.com/vhvb1989/azure-sdk-for-cpp/blob/moving-json-to-internal/sdk/core/azure-core/inc/azure/core/internal/json.hpp#L123-L2703

**Comment by vhvb1989** [comment]
> 
> 
> The diff is not visible because the file is too large, but we're adding ~2000 lines worth of macro definitions alone, none of them are AZ_ prefixed, most can be removed, and all would collide with NLOHMANN_JSON. Please remove most of them that can be removed, AZ_ prefix the rest.

All definitions are unset at the end of the file

**Comment by vhvb1989** [comment]
I will create a sample app that uses the SDK and also some other version of nlohmann lib and make sure it buils and runs :)

**Comment by antkmsft** [comment]
@vhvb1989 , I still can see a few defined, `NLOHMANN_JSON_SERIALIZE_ENUM`, `INCLUDE_NLOHMANN_JSON_FWD_HPP_`, there could be more. Can you use some regexp to generate the full list and make sure we're not missing any?

**Comment by vhvb1989** [comment]
> `NLOHMANN_JSON_SERIALIZE_ENUM`

Sure. I think those would come up when trying to build some app that is already using the nlohmann lib.  I will try it now. Thank you

**Comment by antkmsft** [comment]
@vhvb1989 , after thinking more, `#undef` approach is still going to interfere with customer code using nlohmann json. We still need to `AZ_`-prefix all the macros. Imagine that I included nlohmann.json, and then some Azure header. I want to use nlohmann json's macro, but it is not undeffed!

**Comment by antkmsft** [comment]
For thereference, here is the full list ofmacros defined (sorted alphabetically):
```
INCLUDE_NLOHMANN_JSON_FWD_HPP_
JSON_CATCH
JSON_HAS_CPP_14
JSON_HAS_CPP_17
JSON_HEDLEY_ALWAYS_INLINE
JSON_HEDLEY_ARM_VERSION
JSON_HEDLEY_ARM_VERSION_CHECK
JSON_HEDLEY_ARRAY_PARAM
JSON_HEDLEY_ASSUME
JSON_HEDLEY_BEGIN_C_DECLS
JSON_HEDLEY_C_DECL
JSON_HEDLEY_CLANG_HAS_ATTRIBUTE
JSON_HEDLEY_CLANG_HAS_BUILTIN
JSON_HEDLEY_CLANG_HAS_CPP_ATTRIBUTE
JSON_HEDLEY_CLANG_HAS_DECLSPEC_ATTRIBUTE
JSON_HEDLEY_CLANG_HAS_EXTENSION
JSON_HEDLEY_CLANG_HAS_FEATURE
JSON_HEDLEY_CLANG_HAS_WARNING
JSON_HEDLEY_COMPCERT_VERSION
JSON_HEDLEY_COMPCERT_VERSION_CHECK
JSON_HEDLEY_CONCAT
JSON_HEDLEY_CONCAT_EX
JSON_HEDLEY_CONCAT3
JSON_HEDLEY_CONCAT3_EX
JSON_HEDLEY_CONST
JSON_HEDLEY_CONST_CAST
JSON_HEDLEY_CONSTEXPR
JSON_HEDLEY_CPP_CAST
JSON_HEDLEY_CRAY_VERSION
JSON_HEDLEY_CRAY_VERSION_CHECK
JSON_HEDLEY_DEPRECATED
JSON_HEDLEY_DEPRECATED_FOR
JSON_HEDLEY_DIAGNOSTIC_DISABLE_CAST_QUAL
JSON_HEDLEY_DIAGNOSTIC_DISABLE_CPP98_COMPAT_WRAP_
JSON_HEDLEY_DIAGNOSTIC_DISABLE_DEPRECATED
JSON_HEDLEY_DIAGNOSTIC_DISABLE_UNKNOWN_CPP_ATTRIBUTES
JSON_HEDLEY_DIAGNOSTIC_DISABLE_UNKNOWN_PRAGMAS
JSON_HEDLEY_DIAGNOSTIC_POP
JSON_HEDLEY_DIAGNOSTIC_PUSH
JSON_HEDLEY_DMC_VERSION
JSON_HEDLEY_DMC_VERSION_CHECK
JSON_HEDLEY_EMPTY_BASES
JSON_HEDLEY_EMSCRIPTEN_VERSION
JSON_HEDLEY_EMSCRIPTEN_VERSION_CHECK
JSON_HEDLEY_END_C_DECLS
JSON_HEDLEY_FALL_THROUGH
JSON_HEDLEY_FLAGS
JSON_HEDLEY_FLAGS_CAST
JSON_HEDLEY_GCC_HAS_ATTRIBUTE
JSON_HEDLEY_GCC_HAS_BUILTIN
JSON_HEDLEY_GCC_HAS_CPP_ATTRIBUTE
JSON_HEDLEY_GCC_HAS_DECLSPEC_ATTRIBUTE
JSON_HEDLEY_GCC_HAS_EXTENSION
JSON_HEDLEY_GCC_HAS_FEATURE
JSON_HEDLEY_GCC_HAS_WARNING
JSON_HEDLEY_GCC_NOT_CLANG_VERSION_CHECK
JSON_HEDLEY_GCC_VERSION
JSON_HEDLEY_GCC_VERSION_CHECK
JSON_HEDLEY_GNUC_HAS_ATTRIBUTE
JSON_HEDLEY_GNUC_HAS_BUILTIN
JSON_HEDLEY_GNUC_HAS_CPP_ATTRIBUTE
JSON_HEDLEY_GNUC_HAS_DECLSPEC_ATTRIBUTE
JSON_HEDLEY_GNUC_HAS_EXTENSION
JSON_HEDLEY_GNUC_HAS_FEATURE
JSON_HEDLEY_GNUC_HAS_WARNING
JSON_HEDLEY_GNUC_VERSION
JSON_HEDLEY_GNUC_VERSION_CHECK
JSON_HEDLEY_HAS_ATTRIBUTE
JSON_HEDLEY_HAS_BUILTIN
JSON_HEDLEY_HAS_CPP_ATTRIBUTE
JSON_HEDLEY_HAS_CPP_ATTRIBUTE_NS
JSON_HEDLEY_HAS_DECLSPEC_ATTRIBUTE
JSON_HEDLEY_HAS_EXTENSION
JSON_HEDLEY_HAS_FEATURE
JSON_HEDLEY_HAS_WARNING
JSON_HEDLEY_IAR_VERSION
JSON_HEDLEY_IAR_VERSION_CHECK
JSON_HEDLEY_IBM_VERSION
JSON_HEDLEY_IBM_VERSION_CHECK
JSON_HEDLEY_IMPORT
JSON_HEDLEY_INLINE
JSON_HEDLEY_INTEL_VERSION
JSON_HEDLEY_INTEL_VERSION_CHECK
JSON_HEDLEY_IS_CONSTANT
JSON_HEDLEY_IS_CONSTEXPR_
JSON_HEDLEY_LIKELY
JSON_HEDLEY_MALLOC
JSON_HEDLEY_MESSAGE
JSON_HEDLEY_MSVC_VERSION
JSON_HEDLEY_MSVC_VERSION_CHECK
JSON_HEDLEY_NEVER_INLINE
JSON_HEDLEY_NO_ESCAPE
JSON_HEDLEY_NO_RETURN
JSON_HEDLEY_NO_THROW
JSON_HEDLEY_NON_NULL
JSON_HEDLEY_NULL
JSON_HEDLEY_PELLES_VERSION
JSON_HEDLEY_PELLES_VERSION_CHECK
JSON_HEDLEY_PGI_VERSION
JSON_HEDLEY_PGI_VERSION_CHECK
JSON_HEDLEY_PRAGMA
JSON_HEDLEY_PREDICT
JSON_HEDLEY_PREDICT_FALSE
JSON_HEDLEY_PREDICT_TRUE
JSON_HEDLEY_PRINTF_FORMAT
JSON_HEDLEY_PRIVATE
JSON_HEDLEY_PUBLIC
JSON_HEDLEY_PURE
JSON_HEDLEY_REINTERPRET_CAST
JSON_HEDLEY_REQUIRE
JSON_HEDLEY_REQUIRE_CONSTEXPR
JSON_HEDLEY_REQUIRE_MSG
JSON_HEDLEY_RESTRICT
JSON_HEDLEY_RETURNS_NON_NULL
JSON_HEDLEY_SENTINEL
JSON_HEDLEY_STATIC_ASSERT
JSON_HEDLEY_STATIC_CAST
JSON_HEDLEY_STRINGIFY
JSON_HEDLEY_STRINGIFY_EX
JSON_HEDLEY_SUNPRO_VERSION
JSON_HEDLEY_SUNPRO_VERSION_CHECK
JSON_HEDLEY_TI_ARMCL_VERSION
JSON_HEDLEY_TI_ARMCL_VERSION_CHECK
JSON_HEDLEY_TI_CL2000_VERSION
JSON_HEDLEY_TI_CL2000_VERSION_CHECK
JSON_HEDLEY_TI_CL430_VERSION
JSON_HEDLEY_TI_CL430_VERSION_CHECK
JSON_HEDLEY_TI_CL6X_VERSION
JSON_HEDLEY_TI_CL6X_VERSION_CHECK
JSON_HEDLEY_TI_CL7X_VERSION
JSON_HEDLEY_TI_CL7X_VERSION_CHECK
JSON_HEDLEY_TI_CLPRU_VERSION
JSON_HEDLEY_TI_CLPRU_VERSION_CHECK
JSON_HEDLEY_TI_VERSION
JSON_HEDLEY_TI_VERSION_CHECK
JSON_HEDLEY_TINYC_VERSION
JSON_HEDLEY_TINYC_VERSION_CHECK
JSON_HEDLEY_UNAVAILABLE
JSON_HEDLEY_UNLIKELY
JSON_HEDLEY_UNPREDICTABLE
JSON_HEDLEY_UNREACHABLE
JSON_HEDLEY_UNREACHABLE_RETURN
JSON_HEDLEY_VERSION
JSON_HEDLEY_VERSION_DECODE_MAJOR
JSON_HEDLEY_VERSION_DECODE_MINOR
JSON_HEDLEY_VERSION_DECODE_REVISION
JSON_HEDLEY_VERSION_ENCODE
JSON_HEDLEY_WARN_UNUSED_RESULT
JSON_HEDLEY_WARN_UNUSED_RESULT_MSG
JSON_HEDLEY_WARNING
JSON_INTERNAL_CATCH
JSON_THROW
JSON_TRY
NLOHMANN_BASIC_JSON_TPL
NLOHMANN_BASIC_JSON_TPL_DECLARATION
NLOHMANN_JSON_SERIALIZE_ENUM
```

And here are the ones that we undef:
```
JSON_CATCH
JSON_HEDLEY_ALWAYS_INLINE
JSON_HEDLEY_ARM_VERSION
JSON_HEDLEY_ARM_VERSION_CHECK
JSON_HEDLEY_ARRAY_PARAM
JSON_HEDLEY_ASSUME
JSON_HEDLEY_BEGIN_C_DECLS
JSON_HEDLEY_C_DECL
JSON_HEDLEY_CLANG_HAS_ATTRIBUTE
JSON_HEDLEY_CLANG_HAS_BUILTIN
JSON_HEDLEY_CLANG_HAS_CPP_ATTRIBUTE
JSON_HEDLEY_CLANG_HAS_DECLSPEC_DECLSPEC_ATTRIBUTE
JSON_HEDLEY_CLANG_HAS_EXTENSION
JSON_HEDLEY_CLANG_HAS_FEATURE
JSON_HEDLEY_CLANG_HAS_WARNING
JSON_HEDLEY_COMPCERT_VERSION
JSON_HEDLEY_COMPCERT_VERSION_CHECK
JSON_HEDLEY_CONCAT
JSON_HEDLEY_CONCAT_EX
JSON_HEDLEY_CONCAT3
JSON_HEDLEY_CONCAT3_EX
JSON_HEDLEY_CONST
JSON_HEDLEY_CONST_CAST
JSON_HEDLEY_CONSTEXPR
JSON_HEDLEY_CPP_CAST
JSON_HEDLEY_CRAY_VERSION
JSON_HEDLEY_CRAY_VERSION_CHECK
JSON_HEDLEY_DEPRECATED
JSON_HEDLEY_DEPRECATED_FOR
JSON_HEDLEY_DIAGNOSTIC_DISABLE_CAST_QUAL
JSON_HEDLEY_DIAGNOSTIC_DISABLE_CPP98_COMPAT_WRAP_
JSON_HEDLEY_DIAGNOSTIC_DISABLE_DEPRECATED
JSON_HEDLEY_DIAGNOSTIC_DISABLE_UNKNOWN_CPP_ATTRIBUTES
JSON_HEDLEY_DIAGNOSTIC_DISABLE_UNKNOWN_PRAGMAS
JSON_HEDLEY_DIAGNOSTIC_POP
JSON_HEDLEY_DIAGNOSTIC_PUSH
JSON_HEDLEY_DMC_VERSION
JSON_HEDLEY_DMC_VERSION_CHECK
JSON_HEDLEY_EMPTY_BASES
JSON_HEDLEY_EMSCRIPTEN_VERSION
JSON_HEDLEY_EMSCRIPTEN_VERSION_CHECK
JSON_HEDLEY_END_C_DECLS
JSON_HEDLEY_FALL_THROUGH
JSON_HEDLEY_FLAGS
JSON_HEDLEY_FLAGS_CAST
JSON_HEDLEY_GCC_HAS_ATTRIBUTE
JSON_HEDLEY_GCC_HAS_BUILTIN
JSON_HEDLEY_GCC_HAS_CPP_ATTRIBUTE
JSON_HEDLEY_GCC_HAS_DECLSPEC_ATTRIBUTE
JSON_HEDLEY_GCC_HAS_EXTENSION
JSON_HEDLEY_GCC_HAS_FEATURE
JSON_HEDLEY_GCC_HAS_WARNING
JSON_HEDLEY_GCC_NOT_CLANG_VERSION_CHECK
JSON_HEDLEY_GCC_VERSION
JSON_HEDLEY_GCC_VERSION_CHECK
JSON_HEDLEY_GNUC_HAS_ATTRIBUTE
JSON_HEDLEY_GNUC_HAS_BUILTIN
JSON_HEDLEY_GNUC_HAS_CPP_ATTRIBUTE
JSON_HEDLEY_GNUC_HAS_DECLSPEC_ATTRIBUTE
JSON_HEDLEY_GNUC_HAS_EXTENSION
JSON_HEDLEY_GNUC_HAS_FEATURE
JSON_HEDLEY_GNUC_HAS_WARNING
JSON_HEDLEY_GNUC_VERSION
JSON_HEDLEY_GNUC_VERSION_CHECK
JSON_HEDLEY_HAS_ATTRIBUTE
JSON_HEDLEY_HAS_BUILTIN
JSON_HEDLEY_HAS_CPP_ATTRIBUTE
JSON_HEDLEY_HAS_CPP_ATTRIBUTE_NS
JSON_HEDLEY_HAS_DECLSPEC_ATTRIBUTE
JSON_HEDLEY_HAS_EXTENSION
JSON_HEDLEY_HAS_FEATURE
JSON_HEDLEY_HAS_WARNING
JSON_HEDLEY_IAR_VERSION
JSON_HEDLEY_IAR_VERSION_CHECK
JSON_HEDLEY_IBM_VERSION
JSON_HEDLEY_IBM_VERSION_CHECK
JSON_HEDLEY_IMPORT
JSON_HEDLEY_INLINE
JSON_HEDLEY_INTEL_VERSION
JSON_HEDLEY_INTEL_VERSION_CHECK
JSON_HEDLEY_IS_CONSTANT
JSON_HEDLEY_IS_CONSTEXPR_
JSON_HEDLEY_LIKELY
JSON_HEDLEY_MALLOC
JSON_HEDLEY_MESSAGE
JSON_HEDLEY_MSVC_VERSION
JSON_HEDLEY_MSVC_VERSION_CHECK
JSON_HEDLEY_NEVER_INLINE
JSON_HEDLEY_NO_ESCAPE
JSON_HEDLEY_NO_RETURN
JSON_HEDLEY_NO_THROW
JSON_HEDLEY_NON_NULL
JSON_HEDLEY_NULL
JSON_HEDLEY_PELLES_VERSION
JSON_HEDLEY_PELLES_VERSION_CHECK
JSON_HEDLEY_PGI_VERSION
JSON_HEDLEY_PGI_VERSION_CHECK
JSON_HEDLEY_PREDICT
JSON_HEDLEY_PRINTF_FORMAT
JSON_HEDLEY_PRIVATE
JSON_HEDLEY_PUBLIC
JSON_HEDLEY_PURE
JSON_HEDLEY_REINTERPRET_CAST
JSON_HEDLEY_REQUIRE
JSON_HEDLEY_REQUIRE_CONSTEXPR
JSON_HEDLEY_REQUIRE_MSG
JSON_HEDLEY_RESTRICT
JSON_HEDLEY_RETURNS_NON_NULL
JSON_HEDLEY_SENTINEL
JSON_HEDLEY_STATIC_ASSERT
JSON_HEDLEY_STATIC_CAST
JSON_HEDLEY_STRINGIFY
JSON_HEDLEY_STRINGIFY_EX
JSON_HEDLEY_SUNPRO_VERSION
JSON_HEDLEY_SUNPRO_VERSION_CHECK
JSON_HEDLEY_TI_ARMCL_VERSION
JSON_HEDLEY_TI_ARMCL_VERSION_CHECK
JSON_HEDLEY_TI_CL2000_VERSION
JSON_HEDLEY_TI_CL2000_VERSION_CHECK
JSON_HEDLEY_TI_CL430_VERSION
JSON_HEDLEY_TI_CL430_VERSION_CHECK
JSON_HEDLEY_TI_CL6X_VERSION
JSON_HEDLEY_TI_CL6X_VERSION_CHECK
JSON_HEDLEY_TI_CL7X_VERSION
JSON_HEDLEY_TI_CL7X_VERSION_CHECK
JSON_HEDLEY_TI_CLPRU_VERSION
JSON_HEDLEY_TI_CLPRU_VERSION_CHECK
JSON_HEDLEY_TI_VERSION
JSON_HEDLEY_TI_VERSION_CHECK
JSON_HEDLEY_TINYC_VERSION
JSON_HEDLEY_TINYC_VERSION_CHECK
JSON_HEDLEY_UNAVAILABLE
JSON_HEDLEY_UNLIKELY
JSON_HEDLEY_UNPREDICTABLE
JSON_HEDLEY_UNREACHABLE
JSON_HEDLEY_UNREACHABLE_RETURN
JSON_HEDLEY_VERSION
JSON_HEDLEY_VERSION_DECODE_MAJOR
JSON_HEDLEY_VERSION_DECODE_MINOR
JSON_HEDLEY_VERSION_DECODE_REVISION
JSON_HEDLEY_VERSION_ENCODE
JSON_HEDLEY_WARN_UNUSED_RESULT
JSON_HEDLEY_WARN_UNUSED_RESULT_MSG
JSON_HEDLEY_WARNING
JSON_INTERNAL_CATCH
JSON_THROW
JSON_TRY
```

You can compare two in the diff tool and see which are not undeffed.

**Comment by antkmsft** [comment]
Here is what we've missed, as I see:
* `INCLUDE_NLOHMANN_JSON_FWD_HPP_`
* `JSON_HAS_CPP_14`
* `JSON_HAS_CPP_17`
* `JSON_HEDLEY_CLANG_HAS_DECLSPEC_ATTRIBUTE` (the one we undef is `JSON_HEDLEY_CLANG_HAS_DECLSPEC_DECLSPEC_ATTRIBUTE` which is never defined)
* `JSON_HEDLEY_PRAGMA`
* `JSON_HEDLEY_PREDICT_FALSE`
* `JSON_HEDLEY_PREDICT_TRUE`
* `NLOHMANN_BASIC_JSON_TPL`
* `NLOHMANN_BASIC_JSON_TPL_DECLARATION`
* `NLOHMANN_JSON_SERIALIZE_ENUM`
They still should be `AZ_` (or, `_az_`) prefixed.

**Comment by antkmsft** [comment]
Please double check, I just found out that the tool I was using (Expresso) has ran out of lines because the file is too long, so I might have missed something. But some of them I confirm I can't see being undeffed.

**Comment by vhvb1989** [comment]
> Please double check,

Yes, I have created a simple app that uses the same json lib and it should be compiled with our internal version but it didn't work..

There's more work I need to do to the internal version.  Thank you for your help @antkmsft

**Comment by vhvb1989** [comment]
@antkmsft  I have made the required changes to the internal header so it can be used in one app that is already using json.
See: https://github.com/vhvb1989/azure-sdk-for-cpp/blob/integrate-to-app-using-json/samples/integration/third-party-app-using-same-json-lib/src/main.cpp

The application can use json from Azure internal and from nlohmann :)

**Comment by antkmsft** [comment]
@vhvb1989, but what about prefixing and undeffing the rest of the macros? for example `JSON_HEDLEY_LIKELY`, we define is not prefixed. Please do regexp search and rename for everything deffed, prefix it, and undef in the end.

**Comment by antkmsft** [comment]
@vhvb1989 Here is what I was talking about: https://github.com/vhvb1989/azure-sdk-for-cpp/pull/2
Too bad Gihub can't show diffs in 25k line files still.
Can you verify against your app?

**Review by antkmsft — APPROVED** [review]
